lxd/storage/drivers/drivers_lvm_volumes.go

```diff
@@ -0,0 +1,732 @@
+		err = d.renameLogicalVolume(originalVolDevPath, tmpVolDevPath)
+		if err != nil {
+			return fmt.Errorf("Error temporarily renaming original LVM logical volume: %v", err)
+		}
+
+		revert.Add(func() {
+			// Rename the original volume back to the original name.
+			d.renameLogicalVolume(tmpVolDevPath, originalVolDevPath)
+		})
+
+		// Create writable snapshot from source snapshot named as target volume.
+		_, err = d.createLogicalVolumeSnapshot(d.config["lvm.vg_name"], snapVol, vol, false, true)
+		if err != nil {
+			return fmt.Errorf("Error restoring LVM logical volume snapshot: %v", err)
+		}
+
+		volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, vol.name)
+
+		revert.Add(func() {
+			d.removeLogicalVolume(volDevPath)
+		})
+
+		// If the volume's filesystem needs to have its UUID regenerated to allow mount then do so now.
+		if vol.contentType == ContentTypeFS && renegerateFilesystemUUIDNeeded(d.volumeFilesystem(vol)) {
+			d.logger.Debug("Regenerating filesystem UUID", log.Ctx{"dev": volDevPath, "fs": d.volumeFilesystem(vol)})
+			err = regenerateFilesystemUUID(d.volumeFilesystem(vol), volDevPath)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Finally remove the original logical volume. Should always be the last step to allow revert.
+		err = d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, tmpVolName))
+		if err != nil {
+			return fmt.Errorf("Error removing original LVM logical volume: %v", err)
+		}
+
+		revert.Success()
+		return nil
+	}
+
+	// If the pool uses classic logical volumes, then the process for restoring a snapshot is as follows:
+	// 1. Mount source and target.
+	// 2. Rsync source to target.
+	// 3. Unmount source and target.
+	err = vol.MountTask(func(mountPath string, op *operations.Operation) error {
+		// Copy source to destination (mounting each volume if needed).
+		err = snapVol.MountTask(func(srcMountPath string, op *operations.Operation) error {
+			bwlimit := d.config["rsync.bwlimit"]
+			_, err := rsync.LocalCopy(srcMountPath, mountPath, bwlimit, true)
+			return err
+		}, op)
+		if err != nil {
+			return err
+		}
+
+		// Run EnsureMountPath after mounting and syncing to ensure the mounted directory has the
+		// correct permissions set.
+		err = vol.EnsureMountPath()
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}, op)
+	if err != nil {
+		return fmt.Errorf("Error restoring LVM logical volume snapshot: %v", err)
+	}
+
+	revert.Success()
+	return nil
+}
+
+// DeleteVolume deletes a volume of the storage device. If any snapshots of the volume remain then this function
+// will return an error.
+func (d *lvm) DeleteVolume(vol Volume, op *operations.Operation) error {
+	snapshots, err := d.VolumeSnapshots(vol, op)
+	if err != nil {
+		return err
+	}
+
+	if len(snapshots) > 0 {
+		return fmt.Errorf("Cannot remove a volume that has snapshots")
+	}
```

Comment:
I know we have this pattern in some of the other drivers but I'm not sure it's one we should keep for LVM just as I didn't for ZFS.

The reason I didn't for ZFS is that there may be tooling outside of LXD making backup snapshots, those aren't tracked by LXD, will not be deleted by LXD before deleting a volume, yet we should just get rid of them if the user asked for the volume to go away.

Reply:
@stgraber interesting point, although it raises a different question. Let me explain.

The function `d.VolumeSnapshots` actually `d.vfsVolumeSnapshots` rather than inspecting the logical volumes and the comment it has says:

```
	// We use the vfsVolumeSnapshots rather than inspecting the logical volumes themselves because the origin
	// property of an LVM snapshot can be removed/changed when restoring snapshots, such that they are no
	// marked as origin of the parent volume.
```

The reason for this is that originally I started out using the `lvs` command and then inspecting the `origin` field looking for any volumes that had the current volume as the parent/origin, if it did, then I stripped the parent's prefix from the volume name and returned it as a snapshot name.

However I found that certain operations can destroy the origin property and cause the snapshots to be "unlinked" or linked to a different origin (such as restoring the parent from the snapshot, the restored volume would have the origin set to the original snapshot volume).

So instead I moved to just using the `/var/lib/lxd/storage/container-snapshots/<cname>/` folder to indicate snapshots.

I could alternatively just look at the logical volume name and prefix match on parent volume name and extract the snapshot that way. However I thought this might cause issues if someone created a manual snapshot using the same format that LXD doesn't know about.

At least using only the content's of LXD's container-snapshots directory ensures only snapshots that LXD knows about get deleted.

Happy to change it, but good to discuss the pros/cons of each approach.

Reply:
Ah yeah, using the vfsSnapshots for Snapshots() should ignore any user created snapshots and so should be fine.